### DISCUSSION TITLE
texture on torus

### DISCUSSION THREAD

**Discussion by jobares** [pyvista/pyvista#3469] [Q&A]
Hi all,
I am new here so sorry if my question seems trivial...
I would like to display a specific texture (random pattern) on a torus. I would like the texture to follow the mesh of my torus.
The best I have got so far is doing this:

```py
import numpy as np
import pyvista as pv

N  = 50
l  = 50
RR = 1.5
rr = 1

phi_0 = np.linspace(0, 2 * np.pi, endpoint=True, num=N)
theta_0 = np.linspace(0, 2 * np.pi, endpoint=True, num=l)
phi, theta = np.meshgrid(phi_0, theta_0) 
x = np.cos(theta) * (RR + rr * np.cos(phi))
y = np.sin(theta) * (RR + rr * np.cos(phi))
z = rr * np.sin(phi)

M = np.random.uniform(0,1,(100,100))
tex = pv.numpy_to_texture(M)

curvsurf = pv.StructuredGrid(x, y, z)
curvsurf.texture_map_to_plane(inplace=True)
curvsurf.plot(texture=tex)
```

![Capture d’écran de 2022-10-14 17-34-22](https://user-images.githubusercontent.com/115808178/195885693-3b4957d1-3ee8-4f8b-a3c9-e95061686bda.png)

But in this case I get the vertical projection of my texture on the torus, not the texture wrapping the torus.
Do you see a way to get what I want?
Thanks in advance

**Comment by banesullivan** [comment]
I think your problem is that `texture_map_to_plane()` is the wrong choice for what you want to do. You may have to manually create the texture coordinates to get the image to wrap the surface the way you'd like. 

I'm not totally sure I see how exactly you want the texture to wrap the surface so may need some other examples to understand better.

**Reply by jobares** [reply]
Yes, I do agree texture_map_to_plane() is not what I should use and I should make the texture coordinate I need.
I have all the ingredients since I already have the x, y and z fields mapping the image I used for my texture but I do not know how to make texture coordinate. Could you give me an hint?

**Reply by banesullivan** [reply]
The texture coordinates are a UV mapping of the image coordinates onto your mesh such that every node in the mesh has a UV coordinate corresponding to a point in the texture image. I don't think I'll be able to write up a detailed guide at this time but let me try to point you to some helpful resources:

- A great overview of the concepts of texture mapping: https://en.wikipedia.org/wiki/UV_mapping
- General texture examples in PyVsita: https://docs.pyvista.org/examples/02-plot/texture.html
- Another planar example that might help understand the concepts: https://banesullivan.com/pyvista/examples/geological-map.html

I hope this helps!

**Reply by banesullivan** [reply]
See also:

- https://github.com/pyvista/pyvista-support/issues/230
- https://github.com/pyvista/pyvista-support/issues/168
- https://github.com/pyvista/pyvista-support/issues/257
- https://github.com/pyvista/pyvista-support/issues/4

**Reply by jobares** [reply]
Thanks for pointing me to these directions. It looks like it contains what I need.
Thanks you!
